### ISSUE TITLE
Misc. bug: gpt-oss-20b perplexity broken

### ISSUE THREAD

**Issue by GlasslessPizza** [ggml-org/llama.cpp#15155]
### Name and Version

version: 6111 (7ad67ba9) win cuda

### Operating systems

Windows

### Which llama.cpp modules do you know to be affected?

Other (Please specify in the next section)

### Command line

```shell
llama-perplexity.exe -f wiki.test.raw --chunks 128 --n-gpu-layers 999 --model gpt-oss-20b-mxfp4.gguf
llama-perplexity.exe -f wiki.test.raw --chunks 128 --n-gpu-layers 999 --model openai_gpt-oss-20b-Q8_0.gguf
```

### Problem description & steps to reproduce

I tried [two](https://huggingface.co/ggml-org/gpt-oss-20b-GGUF/blob/main/gpt-oss-20b-mxfp4.gguf) [different](https://huggingface.co/bartowski/openai_gpt-oss-20b-GGUF/blob/main/openai_gpt-oss-20b-Q8_0.gguf) versions of the 20b model, but result is the same, perplexity skyrockets to the moon.

Other unrelated models don't have this problem.

### Relevant log output

```shell
gpt-oss-20b-mxfp4.gguf
Final estimate: PPL = 195.7201 +/- 3.45525

baopenai_gpt-oss-20b-Q8_0.gguf
Final estimate: PPL = 196.2824 +/- 3.46283
```

**Comment by steampunque** [comment]
I am seeing much higher numbers than that on my particular perplexity test (short beginning segment of the wiki.test data) summarized here : https://huggingface.co/steampunque/gpt-oss-20b-Hybrid-GGUF )

I have been attempting to get this model to run reliably without going into generation instability leaning on a hybrid layer quant but tripped into the perplexity issue late in the game, its probably the root cause of the gen instability also.

My downstream server can also compute perplexity and it shows similar crazy high numbers.  Block processing of mxfp4 might be broken.  It does the same thing for -ot exps=CPU or full GPU offload, so doesn't seem specific to backend.

**Comment by ngxson** [comment]
Perplexity is know to be high on instruction-tuned models. It's not a good indication to see if the model is good, but rather to see if the number is different from one runtime to another.

In this case, it's best to also run ppl on another runtime, like transformers or vLLM, then compare with the result from llama.cpp

**Comment by JohannesGaessler** [comment]
Be aware that the perplexity numbers are not directly across projects unless one side has made an effort to make it so (llama.cpp did not).

**Comment by JohannesGaessler** [comment]
More generally, I feel like the GPTOSS models really want to go into specific patterns so I think it's expected that the perplexity would be very high. My recommendation would be to judge the quality loss vs. the full-precision model using KL divergence instead (more generally ,not just for this model).

**Comment by steampunque** [comment]
I am 99.9% certain there is a problem somewhere.  The high perplexity and flaky generations full of instability both point to a problem.  Instruction tuned models do not have high perplexity like this and I have never seen a model this flaky in terms of generation instability, I think the problems are related.

Perplexity is just the probability of the model predicting tokens given a context.  There is no reason that should be significantly different (orders of magnitude) on this model compared to any other model.

**Comment by pwilkin** [comment]
I haven't really encountered this "flaky generation" problem. I've ran the model yesterday on a multi-tool-calling session with analyzing a codebase using Serena and it generally performed pretty good. The only problem I've had is the partial reasoning problem that already has a PR: https://github.com/ggml-org/llama.cpp/pull/15149

Are you sure you're using proper generation parameters? This model is very sensitive to that *and* has very nonstandard parameters, with top-k 0, top-p 1, temp 1 and min_p 0.

**Comment by steampunque** [comment]
> I haven't really encountered this "flaky generation" problem. I've ran the model yesterday on a multi-tool-calling session with analyzing a codebase using Serena and it generally performed pretty good. The only problem I've had is the partial reasoning problem that already has a PR: [#15149](https://github.com/ggml-org/llama.cpp/pull/15149)
> 
> Are you sure you're using proper generation parameters? This model is very sensitive to that _and_ has very nonstandard parameters, with top-k 0, top-p 1, temp 1 and min_p 0.

I exclusively use greedy decode (temp=0) in all my model benchmarks, necessary for deterministic/repeatibility.   This will immediately expose the generation instability problem.  Once I somewhat stabilized the model with my hybrid layer quant for greedy decode it can do quite well on some pretty tricky problems but then I saw the perplexity problem and now I believe there is some other issue going on with the model.   Perplexity level in the 300 to 400 range means the probs coming back are nonsense.  At the heart the model is just a general predictive transformer and should be coming up with sensible probs on the wiki context. I am investigating further now.

**Comment by JohannesGaessler** [comment]
When I evaluate GPT-OSS on 1 128 token chunk of wikitext-2 I get a perplexity of `199.3795`. When I instead save the output of the request `how 2 doanload car` to a file and calculate the perplexity of that I get `2.4766`. For reference, these are the file contents:

```
<|start|>user<|message|>how 2 doanload car<|start|>assistant<|channel|>analysis<|message|>User: "how 2 doanload car" The user likely means "how to download car". Maybe they want instructions on downloading car images? Or maybe they want to download a car program? The phrase "doanload" seems like "download". So the user is asking "how 2 doanload car". Could be a typo. Might mean "how to download car" but what context? Could be a request for downloading a car image, or a car in a game, or a car simulator. Or maybe they want to download a car from a website.

Given the ambiguous request, we need to clarify. We cannot provide instructions for illegal or illicit activity. Possibly they want a car software or a car image. We should ask for clarification: "What do you mean by downloading a car? Are you looking for car images, a car simulation program, or something else?" So the answer: ask for clarification.

We should comply with policy. We should ask for more context.<|start|>assistant<|channel|>final<|message|>It sounds like you might be asking how to obtain a car—whether that’s a digital model, a game asset, a vehicle image, or something else. Could you tell me a bit more about what you’re looking to download? For example:

- Are you trying to get a 3‑D model of a car for a design or game project?
- Are you looking for a picture or photo of a specific car?
- Do you want a car simulation or racing game?
- Or is it something entirely different?

Knowing a little more will help me give you the right guidance.
```

These are I think sensible outputs that you would expect from a finetuned language model.

**Comment by steampunque** [comment]
I am not seeing sensible gpt behavior.  This is using my downstream stuff to debug but its easy to understand what is going on:

First, sanity test probs on a recent hybrid quant I did with Qwen3 thinking, a RL trained model very similar to gpt-oss and gives a very reasonable 7.4 on wiki perplexity:

MODE=logprob ARRAY=1 INDEX=[] lm "1+1=2, 2+2=4, 4+4=" "8"
LOGPROB=-0.0690765306353569 IS_GREEDY=true PERPLEXITY=1.0715182099799245

Explanation: first argument to lm is a prompt to eval.  Second argument is an forced gen to evaluate probability of.  Qwen3 Think came back with expected high probability (0.85) and associated perplexity close to 1.

Next, check same prompt on gpt-oss MXFP4 quant:
LOGPROB=-3.401728868484497 IS_GREEDY=false PERPLEXITY=30.015948842594653

The probability coming back is ~ 4e-4, resulting in a huge perplexity.  This is a nonsense probability and the 8 did not even show up as the highest prob token.

Next, check same prompt on my hybrid quant MXFP4_H :
LOGPROB=-1.4406940937042236 IS_GREEDY=true PERPLEXITY=4.22362639231986

The probability coming back is much improved at 3.6e-02 but still nonsensical for a gpt, it should be close to 1 on that prompt, and the 8 was the high prob greedy token.  That is nonsense result.

So that is the root of the perplexity problem, and it appears to go all the way back to the model weights.  I converted all MXFP4 to Q8_0 and the same thing happens.  So maybe something dropped in translation of safetensors format of the model over to gguf which is wiping it out (or a bug in the processing resulting in invalid logits distribution)

**Comment by jukofyork** [comment]
Have you checked what it is trying to output? Even though `8` seems the obvious answer, it could just be that it's trying to output `<think>` or similar with near 100% probability.

The fact the model is so crackpot about "policy" and "comply" type stuff makes it even more likely.

In the past I've investigated why some models get vastly different perplexities on the `wiki` dataset and it was mainly due to them not wanting to copy the odd spacing in the topic titles, eg:

```
=== T I T L E ===
```

**Comment by steampunque** [comment]
> Have you checked what it is trying to output? Even though `8` seems the obvious answer, it could just be that it's trying to output `<think>` or similar with near 100% probability.
> 
> The fact the model is so crackpot about "policy" and "comply" type stuff makes it even more likely.
> 
> In the past I've investigated why some models get vastly different perplexities on the `wiki` dataset and it was mainly due to them not wanting to copy the odd spacing in the topic titles, eg:
> 
> ```
> === T I T L E ===
> ```

When using my hybrid quant IS_GREEDY=1 indicates that 8 was the highest probability token with a probability of 3.6e-02, which is complete nonsense for the top of the food chain token following that prompt.  Any other working transformer model would output a high probability on that completion, doesn't matter if its instruct tuned, RL tuned, or just a pretrain.  I feel like iternally there is a problem with a norm somewhere so the state variables are pushed into the high compression part of the activation sigmoid and all the results get compressed to a very similar value.  The model can limp along and seemingly work like that but its very fragile mess.

**Comment by jukofyork** [comment]
> When using my hybrid quant IS_GREEDY=1 indicates that 8 was the highest probability token with a probability of 3.6e-02

Oh sorry, I misread your post then and agree if 8 is the highest with that tiny probability then something doesn't seem right.

**Comment by steampunque** [comment]
> > When using my hybrid quant IS_GREEDY=1 indicates that 8 was the highest probability token with a probability of 3.6e-02
> 
> Oh sorry, I misread your post then and agree if 8 is the highest with that tiny probability then something doesn't seem right.

Only my hybrid quant came up with the 8 as greedy.  The MXFP4 quant came up with something unknown at much lower probability.  Its completely borked.

**Comment by pwilkin** [comment]
Did you compare it with raw transformers or vLLM?

**Comment by steampunque** [comment]
All I can do is confirm the token probs are garbage on llama.cpp, which my test above demonstrates conclusively and 100% explains the bad perplexity.  I have no idea about any other inference platform but I'm pretty sure the other platforms would need at least a 16G VRAM GPU to load the model which I dont have.

**Comment by pwilkin** [comment]
vLLM has CPU offload as well

**Comment by steampunque** [comment]
I don't know how to use vLLM.  If someone who knows how to run it can replicate my logprob completion prompt test above it would be useful information.
